### PR TITLE
Solves NPQ-9 (Add Lunge skill)

### PR DESCRIPTION
**Pull request by maxilevi**
This PR will contain the code to add the lunge skill into the game. It's currently a draft so still a WIP.

Here is a [GIF](https://i.imgur.com/jeY68pN.gif) of it in action, let me know what you think and if I am missing something.

### CODE REVIEW

File: scripts/engine/component.py [hunk 1]
```diff
@@ -144,20 +145,56 @@ class Knowledge(RegisteredComponent):
...
 
-    def __init__(self, skills: Dict[str, Dict[str, Any]] = None, skill_order: List[str] = None):
-        if skills is None:
-            skills = {}
-        if skill_order is None:
-            skill_order = []
-
-        self.skill_order = skill_order  # list of skill names, to allow access by index
-        self.skills: Dict[str, Dict[str, Any]] = skills  # skill_name : {Skill, cooldown}
-            # FIXME - how is it str and Skill? Can't be both
```
Comment (maxilevi) [inline scripts/engine/component.py:155]: I felt it was better to add a couple of functions and abstract the usage from the internal implementation. Now if the internal structure of the Knowledge component changes we won't need to change anything outside of that.

File: scripts/nqp/actions/skills.py [hunk 2]
```diff
@@ -223,3 +248,74 @@ def build_effects(self, entity: EntityID) -> List[DamageEffect]:
     def get_animation(self, aesthetic: Aesthetic):
         # we can show animations depending on the direction with self.direction
         return aesthetic.sprites.attack
+
+
+@data_defined_skill
```
Comment (maxilevi) [inline scripts/nqp/actions/skills.py:256]: The decorator `@data_defined_skill` calls the function set_properties on the class so we don't have to repeat the code that loads the properties from the library on every skill.

Comment (maxilevi) [reply scripts/nqp/actions/skills.py:256]: By the way, doing this

```
@function_decorator
class A:
     def __init__():
         pass
```
is equivalent to writing this
```
class A:
     def __init__():
         pass
A = function_decorator(A)
```

File: scripts/engine/component.py [hunk 3]
```diff
@@ -144,20 +145,56 @@ class Knowledge(RegisteredComponent):
...
+    def __init__(self, skills: List[Skill] = None, skill_order: List[str] = None):
+        skills = skills or {}
+        skill_order = skill_order or []
```
Comment (maxilevi) [inline scripts/engine/component.py:150]:
```
if skills_order == None:
    skills_order = []
```
Is the same as
```
skills_order = skills_order or []
```
Why? Basically if `skills_order` is `None` it evaluates to false and we get the second part of the conditional. I felt it was cleaner and less cluttery (more pythonic too and uses less lines!)

This is called the [null coalescing operator](https://en.wikipedia.org/wiki/Null_coalescing_operator), it returns the   parameter that is not null. I usually find it pretty useful.

File: scripts/nqp/actions/skills.py (hunk 2 above)
Comment (Snayff) [reply scripts/nqp/actions/skills.py:256]: That's so slick. I love it.

**Review by Snayff — APPROVED** [review]
Great work, Maximiliano.

**Pull request merged**